e2e/recipes/recipes.go

```diff
@@ -0,0 +1,210 @@
+// SPDX-FileCopyrightText: The RamenDR authors
+// SPDX-License-Identifier: Apache-2.0
+
+package recipes
+
+import (
+	"fmt"
+	"maps"
+
+	recipe "github.com/ramendr/recipe/api/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/ramendr/ramen/e2e/config"
+	"github.com/ramendr/ramen/e2e/types"
+)
+
+const (
+	// defaultHookTimeout is the default timeout in seconds for the resources to come up when executing hook.
+	defaultHookTimeout = 300
+	// Resource is the group type for Resource groups.
+	resource = "resource"
+	// kindRecipe is the Kind value for Recipe resources.
+	kindRecipe = "Recipe"
+	// recipeAPIVersion is the APIVersion value for Recipe resources.
+	recipeAPIVersion = "recipe.ramendr.io/v1alpha1"
+
+	defaultGroupName = "rg1"
+	backup           = "backup"
+	restore          = "restore"
+	group            = "group"
+)
+
+// supportedSelectResources keeps supported select resources.
+var supportedSelectResources = map[string]struct{}{
+	"deployment":  {},
+	"pod":         {},
+	"statefulset": {},
+}
+
+func Generate(ctx types.TestContext, recipeConfig *config.Recipe) (*recipe.Recipe, error) {
+	ctxName := ctx.Name()
+	appNamespace := ctx.AppNamespace()
+
+	hooks, err := generateHooks(ctx, recipeConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate hooks: %w", err)
+	}
+
+	return &recipe.Recipe{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       kindRecipe,
+			APIVersion: recipeAPIVersion,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ctxName,
+			Namespace: appNamespace,
+		},
+		Spec: recipe.RecipeSpec{
+			Hooks:     hooks,
+			Groups:    generateGroups(ctx, appNamespace),
+			Workflows: generateWorkflows(recipeConfig, hooks),
+		},
+	}, nil
+}
+
+func generateHooks(ctx types.TestContext, rc *config.Recipe) ([]*recipe.Hook, error) {
+	var hooks []*recipe.Hook
+
+	if rc.CheckHook {
+		checkHook, err := prepareCheckHook(ctx)
+		if err != nil {
+			return hooks, err
+		}
+
+		hooks = append(hooks, checkHook)
+	}
+
+	if rc.ExecHook {
+		execHook, err := prepareExecHook(ctx)
+		if err != nil {
+			return hooks, err
+		}
+
+		hooks = append(hooks, execHook)
+	}
+
+	return hooks, nil
+}
+
+func prepareCheckHook(ctx types.TestContext) (*recipe.Hook, error) {
+	if checks := ctx.Workload().GetChecks(ctx.AppNamespace()); len(checks) == 0 {
+		return nil, fmt.Errorf("failed to prepare check hooks, no checks defined for workload %q",
+			ctx.Workload().GetName())
+	}
+
+	selectResource := ctx.Workload().GetSelectResource()
+	if _, ok := supportedSelectResources[selectResource]; !ok {
+		return nil, fmt.Errorf("selectResource %q is not supported for check hook in workload %q (choose from %v)",
+			selectResource, ctx.Workload().GetName(), maps.Keys(supportedSelectResources))
```

Comment:
Nice to see modern go! But map.Keys() returns an iterator, so this will log something like:

    (choose from 0x4994c0)

To get slice of keys we need:

```go
slices.Collect(maps.Keys(supportedSelectResources))
```